### ISSUE TITLE
Event Aggregator should have an awaitable publish

### ISSUE THREAD

**Issue by lukas-ais** [Caliburn-Micro/Caliburn.Micro#409]
I've implemented a class using IHandleWithTask.
I found that the EventAggregator does not await the result of the invoked method.
Is there any (design) reason to do so?

The messages can be invoked synchronous or asynchronous by e.g. using PublishOnUIThread and PublishOnUIThreadAsync. So if user want to "fire and forget", so PublishOnUIThreadAsync can be used. Consistently PublishOnUIThread should not return before all Handle calls are complete.

**Comment by yadavvineet** [comment]
Similar case i am facing, i am using "await PublishOnUIThreadAsync()" and have listener as  "async Task Handle(operation message)". in the listener i have a a service call which is invoked using Task Factory. As soon as the flow reaches this line, the parent task where event is published is completed and not awaited.

Similar case can also be reproduced using a simple Task.Delay(3000); in the Handle. As soon as the Handle is called, the control flow returns and publish task is completed.

**Comment by nigel-sampson** [comment]
Not sure how this one slipped through my notice, at the moment it isn't awaited, in part due to not having an async publish.

I'm curious how you'd expect this to work when there are multiple handlers, would they handle in parallel or sequentially?

It's certainly an interesting feature request and I'll update the title to reflect that.

**Comment by lukas-ais** [comment]
@nigel-sampson: There is an async publish: PublishOnUIThreadAsync. Isn't it?
If multiple handlers are in registered with IHandleWithTask, I think they should be processed in parallel, because of individual tasks.

**Comment by nigel-sampson** [comment]
Completely agree.

**Comment by yadavvineet** [comment]
Well, We needed the feature urgently and carried out a workaround. Dont know, but may be it can help in aiding the solution.
From the task where we send the action, we have a Task property in base class which we set in Handles for the result that needs to be awaited. For subsequent multiple, we could chose on completion or somehting.  Then from the place we raise the Publish, we put a manual await on the initial task. This is a workaround that we used.
